Python/fileutils.c

```diff
@@ -1377,14 +1377,13 @@ _Py_fopen(const char *pathname, const char *mode)
         return NULL;
     }
 
-    FILE *f = fopen(pathname, mode);
-    if (f == NULL)
-        return NULL;
-    if (make_non_inheritable(fileno(f)) < 0) {
-        fclose(f);
+    PyObject *path = PyUnicode_DecodeFSDefault(pathname);
```

Comment:
The change should be conditionally defined just for Windows. In the Unix world a path is just a sequence of bytes, with only ASCII slash and null reserved.

Reply:
Agreed with @eryksun. Performing the tests across all OS's wouldn't be necessary, since this is a windows specific issue.

Reply:
Thank you, @eryksun, for the comment. One issue is the fact that `_Py_fopen_obj()` calls `set_inheritable()` with `raise == 1`, but `_Py_fopen()` currently uses `make_non_inheritable()`, which calls `set_inheritable()` with `raise == 0`. I think, for non-Windows, the `make_non_inheritable()` call in `_Py_fopen()` should be replaced with `set_inheritable(fileno(f), 0, 1, NULL)`.

Reply:
It's a problem that `_Py_fopen` doesn't require the GIL to be held, but `_Py_fopen_obj` requires it. It may be better to modify `PyRun_SimpleFileExFlags` instead of tinkering with `_Py_fopen`.

(Another issue is with `pymain_run_startup`, which gets the ANSI encoded `PYTHONSTARTUP` filename via `_Py_GetEnv` and passes it to `_Py_fopen`. This is a problem for users with arbitrary Unicode usernames if the startup file is in their profile directory. In Windows we need to pass the filename from the native wide-chacracter environment to `_Py_wfopen`.)